### DISCUSSION TITLE
New view: function call history

### DISCUSSION THREAD

**Discussion by arihannu** [eclipse-cdt/cdt#165] [Ideas]
Hi CDT devs,

I work at Intel as a software engineer, and I have  a question about contributing a new feature to CDT. So, we have had a long time our own plugin which provides an extra view called "Function Call History". It's a fairly simple view which shows a hierarchical view of every C/C++ function call that was recorded by gdb. We have not been developing it for a long time now, since it's "mature"  and it just ships with every Intel oneAPI Base Toolkit release, and is optionally installed to user's Eclipse installation. Now, we are thinking about upstreaming this view to CDT. I'm curious how the process would go and how long this might take to get this new view accepted as part of CDT? I know it's to easy to say, since the review might take a long time and there are also some restrictions when the view can be used (it requires a specific version of gdb, Linux OS and of course, Intel CPU). So, I'm not even sure if such a feature could be accepted by CDT due to those conditions.

Any comments are welcome. 

Thanks.

**Comment by jonahgraham** [comment]
The view sounds like an interesting view. What would be the maintenance plan going forward? Are there people who know this code still active? Will they be maintaining/testing this contribution once it is in CDT? At the moment I am concerned about bringing additional unmaintained code into CDT.

**Comment by cmorty** [comment]
This sounds quite interesting. But what is the motivation of adding it to CDT, rather then having it as a separate, independent plugin? Having a curated page of cdt-extensions, which is integrated into CDT itself, might be a nice idea, though.

**Comment by ruspl-afed** [comment]
> (it requires a specific version of gdb, Linux OS and of course, Intel CPU)

It definitely requires more details, especially how it is going to became a vendor-neutral functionality. 
Perhaps, you may want to "open source" it first under EPL-2.0 license and then it will be much easier to make it a part of Eclipse CDT or/and other Eclipse projects.

**Reply by jld01** [reply]
Hi @arihannu, can you provide more detail on how this feature is tied to a specific version of GDB, to Linux and to Intel CPUs please? Does it require a patched build of GDB? Does it use Intel CPU-specific code within Linux? Could the feature be generalized to avoid these dependencies?

**Reply by mbooth101** [reply]
Indeed if it is unavoidably Linux-specific it may be a better fit as a contribution to the Linuxtools project: https://github.com/eclipse-linuxtools/

**Comment by arihannu** [comment]
The plugin is already licensed under EPL-2.0. There is also the question of support and documentation, which is lacking quite a lot at the moment from our part. This would have to be fixed if it'd be part of CDT. I am worried about the amount of effort we are facing if upstreamed. The reason why I am asking these questions is that we don't have the resources to test and release this plugin anymore every quarter as part of oneAPI toolkit. But I also understand you don't wish to take it in if it won't have any resources to maintain/support it.

One other option what we are now thinking would be to release it to the Eclipse Marketplace. Then we would take care of publishing it and users would at least find it easier than currently. But there we also have the same issue with the support and maintenance.

Let's see which option our product management will take...

**Reply by ruspl-afed** [reply]
> The plugin is already licensed under EPL-2.0

Great to know this,  @arihannu !
If you can post a link to its published sources, it will answer many of our questions.

**Reply by arihannu** [reply]
> > The plugin is already licensed under EPL-2.0
> 
> Great to know this, @arihannu ! If you can post a link to its published sources, it will answer many of our questions.

Unfortunately the sources are not published online, they are in a zipped file when you install Intel oneAPI Base Toolkit for Linux (debugger/latest/src). This plugin has also some extra features and fixes, which would not be upstreamed.

**Comment by jonahgraham** [comment]
[On the mailing list kesselhaus wrote](https://www.eclipse.org/lists/cdt-dev/msg35384.html):
>
> Hi,
>
> Going to answer here, since I have no github account.
>
> How is this related/compared to the "Call Graph Visualization" in CDT?
> Sounds like doing the same thing.
>
> And, this has the same problem as it sounds like the Intel one .. it
> seems to only run on Linux, not Windows.
> So, I have no clue even about the CDT CGV, since I gave up to use Linux
> 20 years ago out of certain reasons.
>
> Regards,
> kesselhaus

**Reply by mbooth101** [reply]
Do you mean the call graph provided by the Linux Tools project? https://wiki.eclipse.org/Linux_Tools_Project/Callgraph/User_Guide

IIRC this is Linux-specific because it uses SystemTap underneath.
